locations/jclouds/src/main/java/org/apache/brooklyn/location/jclouds/softlayer/SoftLayerSameVlanLocationCustomizer.java

```diff
@@ -0,0 +1,385 @@
+        // Check we are customising a SoftLayer location
+        String provider = location.getProvider();
+        if (!(provider.equals(SoftLayerConstants.SOFTLAYER_PROVIDER_NAME))) {
+            String message = String.format("Invalid location provider: %s", provider);
+            LOG.warn(message);
+            throw new IllegalArgumentException(message);
+        }
+
+        // Lookup the latch for this scope
+        String scopeUid = getScopeUid(location);
+        CountDownLatch latch = null;
+        synchronized (lock) {
+            latch = lookupCountDownLatch(location, scopeUid);
+            if (latch == null) {
+                throw new IllegalStateException("No latch available for scope: " + scopeUid);
+            }
+        }
+
+        try {
+            // Check if VLAN IDs have been set for this scope
+            LOG.debug("Looking up saved VLAN details {}", scopeUid);
+            Integer publicVlanId = lookupPublicVlanId(location, scopeUid);
+            Integer privateVlanId = lookupPrivateVlanId(location, scopeUid);
+            if (privateVlanId != null && publicVlanId != null) {
+                LOG.debug("SoftLayer VLANs private {} and public {} already configured for scope: {}",
+                        new Object[] { privateVlanId, publicVlanId, scopeUid });
+                saveVlanDetails(machine, scopeUid, privateVlanId, publicVlanId);
+                return;
+            }
+
+            // No VLAN info yet known, we are probably the first VM and we should
+            // set the VLAN info for others to then learn about.
+
+            // Ask SoftLayer API for the VLAN details for this VM
+            LOG.debug("Requesting VLAN details from API for scope: {}", scopeUid);
+            VirtualGuestApi api = computeService.getContext().unwrapApi(SoftLayerApi.class).getVirtualGuestApi();
+            Long serverId = Long.parseLong(machine.getJcloudsId());
+            VirtualGuest guest = api.getVirtualGuestFiltered(serverId,
+                    "primaryNetworkComponent;" +
+                    "primaryNetworkComponent.networkVlan;" +
+                    "primaryBackendNetworkComponent;" +
+                    "primaryBackendNetworkComponent.networkVlan");
+            publicVlanId = guest.getPrimaryNetworkComponent().getNetworkVlan().getId();
+            privateVlanId = guest.getPrimaryBackendNetworkComponent().getNetworkVlan().getId();
+
+            // Save the VLAN ids
+            LOG.debug("Saving VLAN details private {} and public {} for scope: {}",
+                    new Object[] { privateVlanId, publicVlanId, scopeUid });
+            savePublicVlanId(location, scopeUid, publicVlanId);
+            savePrivateVlanId(location, scopeUid, privateVlanId);
+            saveVlanDetails(machine, scopeUid, privateVlanId, publicVlanId);
+        } finally {
+            // Release the latch
+            latch.countDown();
```

Comment:
Probably no harm though.

We will need to revisit items like this when we federate of course.  (That was another reason for suggesting SemaphoreWithOwner as we own that class, we could swap it for a federated backing store and this would work.  But we can address that when we do proper federation, unless of course the improved SL networking arrives first in which case this becomes redundant!)